cdnauth/auth.py

```
@@ -0,0 +1,119 @@
+        json_settings["debug"] = current_app.config["DEBUG"]
+        json_settings["sp"]["entityId"] = current_app.config["SP_ENTITY_ID"]
+        json_settings["sp"]["assertionConsumerService"]["url"] = current_app.config[
+            "SP_ACS_URL"
+        ]
+        json_settings["sp"]["x509cert"] = current_app.config["SP_CERT"]
+        json_settings["sp"]["privateKey"] = current_app.config["SP_KEY"]
+        json_settings["idp"]["entityId"] = current_app.config["IDP_ENTITY_ID"]
+        json_settings["idp"]["singleSignOnService"]["url"] = current_app.config[
+            "IDP_SSO_URL"
+        ]
+        json_settings["idp"]["x509cert"] = current_app.config["IDP_CERT"]
+        json_settings["security"]["wantAssertionsEncrypted"] = current_app.config[
+            "SP_SECURITY_ASSERTIONS_ENCRYPTED"
+        ]
+
+    return json_settings
+
+
+def prepare_flask_request(request):
+    url_data = urlparse(request.url)
+    return {
+        "https": "on" if request.scheme == "https" else "off",
+        "http_host": request.host,
+        "server_port": url_data.port,
+        "script_name": request.path,
+        "get_data": request.args.copy(),
+        "post_data": request.form.copy(),
+    }
+
+
+bp = Blueprint("auth", __name__, url_prefix="/saml")
+
+
+@bp.route("/", methods=("GET", "POST"))
+def saml():
+    saml_settings = load_saml_settings()
+    req = prepare_flask_request(request)
+    auth = OneLogin_Saml2_Auth(req, saml_settings)
+    errors = []
+    next_page = request.args.get("next")
+    if not next_page or is_safe_url(next_page) is False:
+        next_page = ""
+
+    if "sso" in request.args:
+        return redirect(auth.login(return_to=next_page))
+
+    elif "acs" in request.args:
+        auth.process_response()
+        errors = auth.get_errors()
+        if not auth.is_authenticated():
+            # TODO: return something helpful to the user.
+            # That said, this should never happen.
+            pass
```

Is this because the lambda edge before this would ensure that any requests that make it this far are authenticated?  Or am I thinking about this wrong?

I'd propose one of two things given this `TODO`:
1. a docstring for the route as a whole, explaining this situation where a non-authenticated request makes it this far
2. maybe raise an explicit exception here making it clear to those just reading code that this state really shouldn't be encountered

---

I non-authenticated request should _never_ get this far because Touchstone does not send non-authenticated users back here... if they fail authentication they stay in Touchstone... and if they are _not_ authenticated we send them _to_ Touchstone. This code block is from the example python3-saml provides so I suspect it was just to show what is going on more than making sense to actually do anything. I'll try to make it a lot clearer in the code what is going on. I suspect maybe just logging an exception if this block ever gets run so we'll know I was completely wrong and we need to write code or not :)